analysea/tide.py

```diff
@@ -113,86 +102,107 @@ def demean_amps_phases(
+    ts = df.index
+    df = h_rsmp.iloc[:, 0].values
+    coef = utide.solve(ts, df, lat=lat, **kwargs)
     return coef
 
 
-def detide(
-    ts: pd.Series[float],
-    kwargs: Dict[str, Any],
-    constituents: dict[str, float] | None = None,
+def reconstruct_chunk(
+    ts_chunk: pd.Series, constituents: Dict[Any, Any], **kwargs: Optional[Dict[Any, Any]]
 ) -> pd.Series:
-    if constituents is None:
-        constituents = calc_constituents(ts=ts, kwargs=kwargs)
-        tidal = utide.reconstruct(ts.index, nd_format(constituents), verbose=kwargs["verbose"])
-    else:
-        tidal = utide.reconstruct(ts.index, nd_format(constituents), verbose=kwargs["verbose"])
-    storm_surge = ts - tidal.h
-    return storm_surge
+    """
+    Reconstruct a single chunk of time series data.
+    This function is mainly aimed to reduce intense i/o loads (RAM usage)
+    """
 
+    tidal = utide.reconstruct(ts_chunk.index, nd_format(constituents), **kwargs)
 
-def detide_yearly(h: pd.Series[float], split_period: int, kwargs: Dict[str, Any] = OPTS) -> pd.Series:
-    log = kwargs.get("verbose", False)
+    return pd.Series(data=ts_chunk.iloc[:, 0].values - tidal.h, index=ts_chunk.index)
 
-    min_time = pd.Timestamp(h.index.min())
-    max_time = pd.Timestamp(h.index.max())
-    date_ranges = pd.date_range(start=min_time, end=max_time, freq=f"{split_period}D")
-    surge = pd.DataFrame([])
 
-    for start, end in zip(date_ranges[:-1], date_ranges[1:]):
-        signal = h[start:end]
-        surge_tmp = detide(signal, kwargs)
-        if not surge_tmp.empty:
-            surge = pd.concat([surge, surge_tmp])
-        if log:
-            print(f"  => Analyse year {start.year} ({start}-{end})")
-            print(f"   +>  {len(surge)} / {len(h)} records done")
+def detide(
+    ts: pd.Series[float],
+    constituents: Dict[Any, Any] | None = None,
+    lat: float = 0,
+    resample_time: int = 10,
+    split_period: int = 365,
+    **kwargs: Optional[Dict[Any, Any]],
+) -> pd.Series:
+    """
+    By default this function will split the time series into yearly chunks
+
+    @param ts: time series to be processed
+    @param constituents: constituents to be used in the analysis
+    @param lat: latitude of the station
+    @param resample_time: resample time in minutes
+    @param split_period: period in days to split the time series into (default 365)
+    @param kwargs: keyword arguments to be passed to utide.reconstruct
+    @return: reconstructed time series
+    """
+    result_series = []
+    date_ranges = pd.date_range(start=ts.index[0], end=ts.index[-1], freq=f"{split_period}D")
+    # Split the time series into 6-month chunks (default), otherwise specify
+    for start_date, end_date in zip(date_ranges[:-1], date_ranges[1:]):
+        ts_chunk = ts[start_date:end_date]
+        if constituents is None:
+            constituents = calc_constituents(ts=ts, resample_time=resample_time, lat=lat, **kwargs)
```

Comment:
Shouldn't the constituents calculation be outside of the loop?

Reply:
moved out of loop